### ISSUE TITLE
blinking caret initially not visible

### ISSUE THREAD

**Issue by hwaite** [TomasMikula/RichTextFX#189]
Upon receiving focus, `CodeArea` does not immediately display cursor. It seems that `StyledTextAreaVisual.caretVisible`'s initial state is false? Usability would be enhanced if caret were displayed faster. Otherwise, user is left wondering which control is truly active.

**Comment by TomasMikula** [comment]
You are right. When the area does not have focus, the value of `caretVisible` is `false`. When it receives focus, `caretVisible` switches to alternating every 500ms between `false` and `true`. There are actually two problems currently: 1. the alternating starts only after the first 500ms; and 2. the alternating starts at `false`. As a result, it takes 1 second until the caret is first displayed.

It is trivial to fix 2. (all it takes is to change one boolean value in the source code), getting the visual feedback down to 500ms. Fixing 1. is less trivial to do without increasing the complexity of the code. The cleanest solution would be if there was a version of ReactFX's [EventStreams.ticks(Duration)](http://www.reactfx.org/javadoc/2.0-M4u1/org/reactfx/EventStreams.html#ticks-java.time.Duration-) that emitted the first tick immediately.

**Comment by hwaite** [comment]
It would also be nice if timer were reset with each keystroke. If user types quickly enough, caret should never disappear.

**Comment by TomasMikula** [comment]
You are right again.

**Comment by JordanMartinez** [comment]
Those EventStreams have now been written. A new ReactFX release needs to be made. Then we'd need to update the gradle file to use that new release. Then we can update the code in the (now skinless) StyledTextArea.

**Comment by JordanMartinez** [comment]
So... I'm having trouble implementing this feature... 

One issue is that the `AccumulatingEventStream` returned from `accumulate()` needs to be "restarted" when ticks gets restarted:
````java
EventStream<?> ticks = restartableTicks0(duration, keyTyped);
ticks.accumulate(false, (b, x) -> !b);
````
After so many seconds, the events emitted so far are: "true, false, true, false, true." The last event in this sequence is `true`. Now, let's say the user types a key at this point. `ticks` will restart itself and immediately emit a tick. `accumulate()` will then return `false` (do not show the caret). 
If it could reset itself upon an impulse stream's event, so that the initial value was `false` again, then it would return `true` (show the caret) every time the user types a key. Since `ticks`' timer restarts, the next value (false) wouldn't be emitted until  `duration` amount of time passes (simulating the caret blink).

Although I'm very sure there's a better way to write this (or maybe a new EventStream class that needs to be written), I was able to get this to work with the following code (ugly in my opinion):
````java
// caret visibility code:
EventStream<KeyEvent> keyTyped = EventStreams.eventsOf(this, KeyEvent.KEY_TYPED);

caretVisible = EventStreams.valuesOf(blinkCaret)
    .flatMap(blink -> blink
        ? booleanPulse(Duration.ofMillis(500), keyTyped)
        : EventStreams.valuesOf(Val.constant(false)))
                .toBinding(false);

// boolean pulse code
private static EventStream<Boolean> booleanPulse(Duration duration, EventStream<?> impulse) {
    EventStream<?> ticks = EventStreams.restartableTicks0(duration, impulse);
    EventStream<Boolean> forceCaretVisiblity = impulse.map(x -> true);

    EventStream<?> pulses = EventStreams.merge(ticks, forceCaretVisiblity);

    // pulses will emit 2 events when key is typed.
    // the first returns false (setting up the next one)
    // the second one (null) will invert the original value to true.
    return pulses.accumulate(false, (b, x) -> {
       if (x != null) {
           System.out.println("returning false");
           return false;
       } else {
           System.out.println("returning " + !b);
           return !b;
       }
    });
}
````

Another issue relates to the impulse that is used to restart ticks. 
When I use keyTyped, then the caret will continue to show when I press and hold a regular key (letter, digit, Backspace). 
However, if I press and hold an arrow key to navigate throughout the area, the stream will only emit one KeyEvent as soon as I press it and no more after that. Thus, the caret will return to its blinking state rather than an always-on state as I navigate through the area.

**Comment by TomasMikula** [comment]
So in dd5eda7 I did kind of the same thing as you, but using a _state machine_ combinator, with two input streams, typing and ticks, causing state transitions.

**Comment by JordanMartinez** [comment]
Huh... I didn't even know about the `StateMachine` class...  
Cool, it's done.

**Comment by TomasMikula** [comment]
Yeah, this one was easier to explain in code than in words.

One thing worth noting is that there is already an event stream at hand that emits an event when caret or text has been manipulated, namely `caretDirty`, so that makes your problem with keyTyped go away.

And then the multi-input state machine (which is basically a generalization of `accumulate` to multiple inputs) lets you define the state transition for each input separately, so you don't have to merge two streams into one and then in accumulate reconstruct the information which input stream the event came from. I had a [short blog post](http://tomasmikula.github.io/blog/2014/05/01/reactfxs-general-stream-combinator-state-machine.html) about the state machine, but there is not much more to it than you can already see from this example.

The snippet you posted depends on ReactFX implementation details in two ways:
1. It assumes that `restartableTicks0` emits `null`s. Which it in fact does, but the type `EventStream<?>` does not say anything about the type of objects it emits.
2. It assumes that when `restartableTicks0` stream is restarted, the initial tick is not emitted immediately from within the restart impulse handler, but at some `ε` later, which again is the case because of how we (i.e. you ;)) implemented `restartableTicks0` (the initial tick, although at timeframe 0, is still emitted asynchronously from an animation), but that fact is not part of the "specification".

**Comment by JordanMartinez** [comment]
Yeah, that's a better way to write it then.
The blog post explained the terminology I otherwise didn't understand. Thanks for posting that.

**Comment by JordanMartinez** [comment]
Just a heads-up. I compiled master and was using it in my code when this exception was thrown. I was traversing through some TextFields via the `Tab` key before clicking on a blank line (line w/o any text) in the area:
````java
Exception in thread "JavaFX Application Thread" java.lang.NoSuchMethodError: org.reactfx.EventStreams.restartableTicks(Ljava/time/Duration;Lorg/reactfx/EventStream;)Lorg/reactfx/EventStream;
	at org.fxmisc.richtext.StyledTextArea.booleanPulse(StyledTextArea.java:1243)
	at org.fxmisc.richtext.StyledTextArea.lambda$new$9(StyledTextArea.java:599)
	at org.reactfx.FlatMapStream.lambda$observeInputs$189(FlatMapStream.java:23)
	at org.reactfx.util.NonAccumulativeStreamNotifications.lambda$head$272(NotificationAccumulator.java:134)
	at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:68)
	at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:57)
	at org.reactfx.ProperEventStream.emit(ProperEventStream.java:18)
	at org.reactfx.EventStreams$3.lambda$observeInputs$7(EventStreams.java:104)
	at com.sun.javafx.binding.ExpressionHelper$SingleChange.fireValueChangedEvent(ExpressionHelper.java:182)
	at com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:81)
	at javafx.beans.binding.BooleanBinding.invalidate(BooleanBinding.java:176)
	at javafx.beans.binding.Bindings$ShortCircuitAndInvalidator.invalidated(Bindings.java:4555)
	at com.sun.javafx.binding.ExpressionHelper$SingleInvalidation.fireValueChangedEvent(ExpressionHelper.java:137)
	at com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:81)
	at javafx.beans.binding.BooleanBinding.invalidate(BooleanBinding.java:176)
	at javafx.beans.binding.Bindings$ShortCircuitAndInvalidator.invalidated(Bindings.java:4555)
	at com.sun.javafx.binding.ExpressionHelper$SingleInvalidation.fireValueChangedEvent(ExpressionHelper.java:137)
	at com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:81)
	at javafx.beans.property.ReadOnlyBooleanPropertyBase.fireValueChangedEvent(ReadOnlyBooleanPropertyBase.java:72)
	at javafx.scene.Node$FocusedProperty.notifyListeners(Node.java:7718)
	at javafx.scene.Scene$13.invalidated(Scene.java:2080)
	at javafx.beans.property.ObjectPropertyBase.markInvalid(ObjectPropertyBase.java:111)
	at javafx.beans.property.ObjectPropertyBase.set(ObjectPropertyBase.java:146)
	at javafx.scene.Scene$KeyHandler.setFocusOwner(Scene.java:3924)
	at javafx.scene.Scene$KeyHandler.requestFocus(Scene.java:3971)
	at javafx.scene.Scene$KeyHandler.access$1900(Scene.java:3910)
	at javafx.scene.Scene.requestFocus(Scene.java:2044)
	at javafx.scene.Node.requestFocus(Node.java:7879)
	at org.fxmisc.richtext.StyledTextAreaBehavior.mousePressed(StyledTextAreaBehavior.java:439)
	at org.fxmisc.wellbehaved.event.EventHandlerTemplate$On.lambda$null$23(EventHandlerTemplate.java:156)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.fxmisc.wellbehaved.event.EventHandlerTemplate$On.lambda$act$24(EventHandlerTemplate.java:155)
	at org.fxmisc.wellbehaved.event.EventHandlerTemplate$Builder.lambda$create$21(EventHandlerTemplate.java:117)
	at org.fxmisc.wellbehaved.event.EventHandlerTemplate.lambda$bind$14(EventHandlerTemplate.java:18)
	at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
	at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
	at javafx.event.Event.fireEvent(Event.java:198)
	at javafx.scene.Scene$MouseHandler.process(Scene.java:3757)
	at javafx.scene.Scene$MouseHandler.access$1500(Scene.java:3485)
	at javafx.scene.Scene.impl_processMouseEvent(Scene.java:1762)
	at javafx.scene.Scene$ScenePeerListener.mouseEvent(Scene.java:2494)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:352)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:275)
	at java.security.AccessController.doPrivileged(Native Method)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleMouseEvent$354(GlassViewEventHandler.java:388)
	at com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:389)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.handleMouseEvent(GlassViewEventHandler.java:387)
	at com.sun.glass.ui.View.handleMouseEvent(View.java:555)
	at com.sun.glass.ui.View.notifyMouse(View.java:937)
	at com.sun.glass.ui.gtk.GtkApplication._runLoop(Native Method)
	at com.sun.glass.ui.gtk.GtkApplication.lambda$null$49(GtkApplication.java:139)
	at java.lang.Thread.run(Thread.java:745)
````

**Comment by JordanMartinez** [comment]
Here's an update. When I created a VBox with a TextField and an InlineCssTextArea and followed the steps outlined above, the exception wasn't thrown. The problem seems to be with something in my code. If I start my program and just click on the area, I get the above exception thrown.

**Comment by JordanMartinez** [comment]
Even after updating the artifact and refreshing my gradle project, I still get that exception... Weird.
